### PR TITLE
Improve self-hosted sites login 

### LINKED ISSUE #3805: Update the "internals" of .ORG login flow
The login flow for .org sites should look more like the diagrams below.
main
discovery
validation
bad bahavior plugins

This ticket changes the internals of .org sites login. Among other changes, it tries the user inserted URL at the beginning of the process, and only later, if necessary, the "discover process" is started.  
These improvements are the companions of the HealthCheck tool, and diagrams above are subject to changes along the way.

### PR DESCRIPTION
**Pull request by daniloercoli**
Fix #3805 by providing an enhanced version of the XML-RPC discovery procedure and some other nice things ;)

- This PR also checks that the found endpoint does have the basic XML-RPC methods required by the app. (We may want to extended the list if necessary).

- The Bad Behavior plugins matching is ready, but moved in the `future` folder. If you feel confident in keeping the BB plugins list updated (wordpress-mobile/app-blocking-plugins) feel free to put it back in production.
The Bad Behavior matching seems to be solid, since the "main" logic behind it was taken from wpscan, but I didn't test it too much. In the end it is just a series of tests upon HTTP response codes returned by the external web server.

- Also added a new Analytics event LOGIN_INSERTED_INVALID_URL (`login_inserted_invalid_url`), that has 1 property attached to it `user_inserted_url`, that keep tracks of how many invalid URLs we receive in the login screen. *I've not added it to Mixpanel* Should I? 

cc @oguzkocer @hypest @SergioEstevao @bummytime @astralbodies

### CODE REVIEW

**Comment by hypest** [comment]
I'm planning to extend the work done here by extracting the cross-platform flows into testable chunks.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPOrg.java [hunk 1]
```diff
@@ -71,154 +78,301 @@ public void fetchBlogList(Callback callback) {
...
-        String rsdUrl;
-        rsdUrl = ApiHelper.getRSDMetaTagHrefRegEx(baseUrl);
-        if (rsdUrl == null) {
-            rsdUrl = ApiHelper.getRSDMetaTagHref(baseUrl);
-        }
-        return rsdUrl;
-    }
-
     private boolean isHTTPAuthErrorMessage(Exception e) {
         if (e != null && e.getMessage() != null && e.getMessage().contains("401")) {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPOrg.java]: Maybe we can just return this since we are not doing anything else in the `if` statement anymore.

Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPOrg.java]: Right.

File: WordPress/src/future/res/PluginsCheckerWPOrg.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        } catch (JSONException e) {
+            AppLog.e(AppLog.T.NUX, "Error while parsing the list of plugins returned from the server.", e);
+            return null;
+        }
+
+        // we have the list. Start the process of checking for plugins
+        String baseURL = getBaseURL(mOriginalURL);
+
+        if (!baseURL.contains("/plugins")) {
+            baseURL = baseURL + "/wp-content/plugins/";
```
Comment (oguzkocer) [inline WordPress/src/future/res/PluginsCheckerWPOrg.java]: This might be a stupid question as I don't know how this class is going to be used, but is there any way the url contains `/wp-content` but not `/plugins`?

Comment (oguzkocer) [inline WordPress/src/future/res/PluginsCheckerWPOrg.java]: Please ignore the above question, I found my answer in `getBaseUrl`, sorry about that!

File: WordPress/src/future/res/PluginsCheckerWPOrg.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+        AppLog.i(AppLog.T.NUX, "The calculated plugins URL is the following: " + baseURL);
+
+        if (!URLUtil.isValidUrl(baseURL)) {
+            AppLog.w(AppLog.T.NUX, "The calculated plugins URL isn't a valid URL. Returning now.");
+            return null;
+        }
+
+        int respCode = openConnection(baseURL);
+        if (respCode != HttpURLConnection.HTTP_OK && respCode != 401 && respCode != 403) {
```
Comment (oguzkocer) [inline WordPress/src/future/res/PluginsCheckerWPOrg.java]: I am a bit confused here, what do we do if the error code is `401` or `403` for example? These cases don't seem to be handled later.

Comment (oguzkocer) [inline WordPress/src/future/res/PluginsCheckerWPOrg.java]: OK, I have an idea. Is this just checking if we can access the server even if it's an unauthorized connection? If so, then this makes sense. However, wouldn't that cause the next call we are going to make to check if a plugin exists return error as well? I really don't know much about how self-hosted installations work, so I might be way off in my assumptions.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPOrg.java [hunk 4]
```diff
@@ -71,154 +78,298 @@ public void fetchBlogList(Callback callback) {
...
+                "wp.getOptions", "wp.uploadFile", "wp.newCategory",
+                "wp.getTags", "wp.getCategories", "wp.editPage", "wp.deletePage",
+                "wp.newPage", "wp.getPages" };
+
+        for (String currentRequiredMethod: requiredMethods) {
+            boolean match = false;
+            for (Object currentAvailableMethod: availableMethods) {
+                if ((currentAvailableMethod).equals(currentRequiredMethod)) {
+                    match = true;
+                    continue;
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPOrg.java]: Getting a compiler warning here: `"continue" is unnecessary as the last statement in a loop`

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPOrg.java [hunk 5]
```diff
@@ -71,154 +78,298 @@ public void fetchBlogList(Callback callback) {
...
+    }
+
+    // Append "xmlrpc.php" if missing in the URL
+    private String appendXMLRPCPath(String url) throws IllegalArgumentException {
+        if (!URLUtil.isValidUrl(url)) {
+            throw new IllegalArgumentException("Input URL " + url + " is not valid!");
+        }
+
+        if (!url.contains("xmlrpc.php")) { // Do not use 'ends' here! Some hosting wants parameters passed to baseURL/xmlrpc-php?my-authcode=XXX
+            if (url.substring(url.length() - 1, url.length()).equals("/")) {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPOrg.java]: Very minor nitpick: Using `charAt` might makes this a bit more readable:

`if (url.charAt(url.length() - 1) == '/')`

File: WordPress/src/future/res/PluginsCheckerWPOrg.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        } catch (IllegalArgumentException e) {
+            AppLog.e(AppLog.T.NUX, "Can't clean the original url: " + sanitizedURL, e);
+        }
+        while (sanitizedURL.endsWith("/")) {
+            sanitizedURL = sanitizedURL.substring(0, sanitizedURL.length() - 1);
+        }
+        return sanitizedURL;
+    }
+
+    private String truncateURLAtPrefix(String url, String prefix) throws IllegalArgumentException {
```
Comment (oguzkocer) [inline WordPress/src/future/res/PluginsCheckerWPOrg.java]: It looks like we are using this same method in `FetchBlogListWPOrg`, should we move this to an util class?

**Comment by oguzkocer** [comment]
@daniloercoli I've completed my review and left a few comments/questions. This is a complicated PR for me to take alone, so getting another set of eyes (more experienced ones) would be good.

Are there any testing instructions I can follow?

File: WordPress/src/future/res/PluginsCheckerWPOrg.java (hunk 3 above)
Comment (daniloercoli) [inline WordPress/src/future/res/PluginsCheckerWPOrg.java]: We need to check wether the `http://domain/wp-content/plugins/` address is available on internet or not. If the HTTP response code in  [200, 401, 403] means that the folder is available on the server. If the response is != from one of the codes above, we can safely assume that the folder is not there. 
Note that the routine is just an heuristic, and we can't be 100% sure the folder is there.

Comment (daniloercoli) [inline WordPress/src/future/res/PluginsCheckerWPOrg.java]: note2: Assuming that the folder is NOT there is not a problem, since the plugins check is just skipped. This minimize false positive.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPOrg.java (hunk 4 above)
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPOrg.java]: Right. It should be `break;`.

File: WordPress/src/future/res/PluginsCheckerWPOrg.java (hunk 6 above)
Comment (daniloercoli) [inline WordPress/src/future/res/PluginsCheckerWPOrg.java]: Yes, I noticed this, but `PluginsCheckerWPOrg` is a work in progress, not sure that the two methods will end up to be equals. Probably not. Adding just a note here, since PluginsCheckerWPOrg will be changed in the future.

Comment (daniloercoli) [inline WordPress/src/future/res/PluginsCheckerWPOrg.java]: Btw, the two functions are already not exactly the same.

**Comment by daniloercoli** [comment]
@oguzkocer I fixed the issues you've brought up during review.
We've a list of self-hosted sites, might make sense to try the new login on them. 
I've tested the patch locally, by messing up with a local installation of WP.

Hint for reviewers: The best way to review all the login flow is to start looking at `FetchBlogListWPOrg.getSelfHostedXmlrpcUrl` and follow the flow from there. Looking at the diff could be hard to follow the new logic.

File: WordPress/src/future/res/PluginsCheckerWPOrg.java (hunk 3 above)
Comment (oguzkocer) [inline WordPress/src/future/res/PluginsCheckerWPOrg.java]: Thanks for the clarification. It's pretty convoluted.

File: WordPress/src/future/res/PluginsCheckerWPOrg.java (hunk 6 above)
Comment (oguzkocer) [inline WordPress/src/future/res/PluginsCheckerWPOrg.java]: Sounds good to me!

**Comment by daniloercoli** [comment]
Closing this PR since #3853 includes the changes added here, and adds Unit Tests.
